File: service/app/retry.py
import time
from dataclasses import dataclass
from typing import Callable, TypeVar

from app.errors import ExtractionValidationError, TransientAPIError

T = TypeVar("T")


@dataclass
class ExtractionOutcome:
    decision: object | None
    failure: str | None  # None | "validation_exhausted" | "transient_exhausted"
    validation_attempts: int
    transient_attempts: int
    last_error: str | None = None


def extract_with_retry_budgets(
    call: Callable[[str | None], T],
    *,
    validation_budget: int,
    transient_budget: int,
    backoff_seconds: float,
    sleep: Callable[[float], None] = time.sleep,
) -> ExtractionOutcome:
    """Run `call(feedback)` with two independent, non-interfering retry budgets.

    `call` takes the previous validation error text (or None on the first/
    non-validation attempts) and returns a parsed, validated result.

    A validation failure feeds its error back to the model on the next
    attempt via `feedback`. A transient failure retries the identical
    request (feedback unchanged) with exponential backoff. Each budget is
    tracked independently per ADR-0008: exhausting one never consumes or is
    blamed on the other.
    """
    feedback: str | None = None
    validation_attempts = 0
    transient_attempts = 0
    last_error: str | None = None

    while True:
        try:
            result = call(feedback)
            return ExtractionOutcome(
                decision=result,
                failure=None,
                validation_attempts=validation_attempts,
                transient_attempts=transient_attempts,
            )
        except ExtractionValidationError as e:
            validation_attempts += 1
            last_error = str(e)
            feedback = last_error
            if validation_attempts > validation_budget:
                return ExtractionOutcome(
                    decision=None,
                    failure="validation_exhausted",
                    validation_attempts=validation_attempts,
                    transient_attempts=transient_attempts,
                    last_error=last_error,
                )
        except TransientAPIError as e:
            transient_attempts += 1
            last_error = str(e)
            if transient_attempts > transient_budget:
                return ExtractionOutcome(
                    decision=None,
                    failure="transient_exhausted",
                    validation_attempts=validation_attempts,
                    transient_attempts=transient_attempts,
                    last_error=last_error,
                )
            sleep(backoff_seconds * (2 ** (transient_attempts - 1)))

File: service/src/reports/retry.ts
/**
 * Two independent, non-interfering retry budgets (ADR-0008). Mirrors
 * app/retry.py's extract_with_retry_budgets() field-for-field.
 *
 * A validation failure feeds its error back to the caller's `call` on the
 * next attempt via `feedback`, so a self-correcting model can fix its own
 * output. A transient failure retries the identical request (feedback
 * unchanged) with exponential backoff. Each budget is tracked independently:
 * exhausting one is never blamed on, or consumes, the other.
 */

import { ExtractionValidationError, TransientAPIError } from './pipeline.errors';

export interface RetryBudgets {
  validation_retry_budget: number;
  transient_retry_budget: number;
  transient_retry_backoff_seconds: number;
}

export type RetryFailure = 'validation_exhausted' | 'transient_exhausted';

export interface RetryOutcome<T> {
  result: T | null;
  failure: RetryFailure | null;
  validationAttempts: number;
  transientAttempts: number;
  lastError: string | null;
}

type Sleep = (seconds: number) => Promise<void>;

const defaultSleep: Sleep = (seconds) => new Promise((resolve) => setTimeout(resolve, seconds * 1000));

interface RetryState {
  feedback: string | null;
  validationAttempts: number;
  transientAttempts: number;
  lastError: string | null;
}

function handleValidationError<T>(
  err: ExtractionValidationError,
  budgets: RetryBudgets,
  state: RetryState,
): RetryOutcome<T> | null {
  state.validationAttempts += 1;
  state.lastError = err.message;
  state.feedback = state.lastError;
  if (state.validationAttempts > budgets.validation_retry_budget) {
    return {
      result: null,
      failure: 'validation_exhausted',
      validationAttempts: state.validationAttempts,
      transientAttempts: state.transientAttempts,
      lastError: state.lastError,
    };
  }
  return null;
}

async function handleTransientError<T>(
  err: TransientAPIError,
  budgets: RetryBudgets,
  state: RetryState,
  sleep: Sleep,
): Promise<RetryOutcome<T> | null> {
  state.transientAttempts += 1;
  state.lastError = err.message;
  if (state.transientAttempts > budgets.transient_retry_budget) {
    return {
      result: null,
      failure: 'transient_exhausted',
      validationAttempts: state.validationAttempts,
      transientAttempts: state.transientAttempts,
      lastError: state.lastError,
    };
  }
  await sleep(budgets.transient_retry_backoff_seconds * 2 ** (state.transientAttempts - 1));
  return null;
}

async function attemptOnce<T>(
  call: (feedback: string | null) => Promise<T>,
  budgets: RetryBudgets,
  state: RetryState,
  sleep: Sleep,
): Promise<RetryOutcome<T> | null> {
  try {
    const result = await call(state.feedback);
    return {
      result,
      failure: null,
      validationAttempts: state.validationAttempts,
      transientAttempts: state.transientAttempts,
      lastError: null,
    };
  } catch (err) {
    if (err instanceof ExtractionValidationError) {
      return handleValidationError<T>(err, budgets, state);
    }
    if (err instanceof TransientAPIError) {
      return handleTransientError<T>(err, budgets, state, sleep);
    }
    throw err;
  }
}

export async function withRetryBudgets<T>(
  call: (feedback: string | null) => Promise<T>,
  budgets: RetryBudgets,
  sleep: Sleep = defaultSleep,
): Promise<RetryOutcome<T>> {
  const state: RetryState = { feedback: null, validationAttempts: 0, transientAttempts: 0, lastError: null };

  for (;;) {
    const outcome = await attemptOnce(call, budgets, state, sleep);
    if (outcome) {
      return outcome;
    }
  }
}
